src/Microsoft.ComponentDetection.Detectors/rust/RustCliDetector.cs

```diff
@@ -0,0 +1,134 @@
+    public override IList<string> SearchPatterns { get; } = new[] { "Cargo.toml" };
+
+    /// <inheritdoc />
+    protected override async Task OnFileFoundAsync(ProcessRequest processRequest, IDictionary<string, string> detectorArgs)
+    {
+        var componentStream = processRequest.ComponentStream;
+
+        try
+        {
+            if (!await this.cliService.CanCommandBeLocatedAsync("cargo", null))
+            {
+                this.Logger.LogWarning("Could not locate cargo command. Skipping Rust CLI detection");
+                return;
+            }
+
+            var cliResult = await this.cliService.ExecuteCommandAsync(
+                "cargo",
+                null,
+                "metadata",
+                "--manifest-path",
+                componentStream.Location,
+                "--format-version=1",
+                "--locked");
+
+            var metadata = CargoMetadata.FromJson(cliResult.StdOut);
+            var graph = BuildGraph(metadata);
+            var root = metadata.Resolve.Root;
+
+            this.TraverseAndRecordComponents(processRequest.SingleFileComponentRecorder, componentStream.Location, graph, root, null, null);
+        }
+        catch (Exception e)
+        {
+            this.Logger.LogWarning(e, "Failed to parse Cargo.toml file: {Location}", processRequest.ComponentStream.Location);
+        }
+    }
+
+    private static Dictionary<string, Node> BuildGraph(CargoMetadata cargoMetadata) => cargoMetadata.Resolve.Nodes.ToDictionary(x => x.Id);
+
+    private static (string Name, string Version) ParseNameAndVersion(string nameAndVersion)
+    {
+        var parts = nameAndVersion.Split(' ');
+        return (parts[0], parts[1]);
+    }
+
+    private void TraverseAndRecordComponents(
+        ISingleFileComponentRecorder recorder,
+        string location,
+        IReadOnlyDictionary<string, Node> graph,
+        string id,
+        DetectedComponent parent,
+        Dep depInfo,
+        bool explicitlyReferencedDependency = false)
+    {
+        try
+        {
+            var isDevelopmentDependency = depInfo?.DepKinds.Any(x => x.Kind is Kind.Dev or Kind.Build) ?? false;
+            var (name, version) = ParseNameAndVersion(id);
+            var detectedComponent = new DetectedComponent(new CargoComponent(name, version));
+
+            recorder.RegisterUsage(
+                detectedComponent,
+                explicitlyReferencedDependency,
+                isDevelopmentDependency: isDevelopmentDependency,
+                parentComponentId: parent?.Component.Id);
+
+            if (!graph.TryGetValue(id, out var node))
+            {
+                this.Logger.LogWarning("Could not find {Id} at {Location} in cargo metadata output", id, location);
+                return;
+            }
+
+            foreach (var dep in node.Deps)
+            {
+                this.TraverseAndRecordComponents(recorder, location, graph, dep.Pkg, detectedComponent, dep, parent == null);
```

Comment:
Not a huge fan of checking for explicit dependencies like this. feedback appreciated

Reply:
I can't think of any other way than `parent == null`. The roots of the graph always require special handling. If I was being nit-picky I would ask you to extract it to a variable to make it explicit `var isDirectDependency = parent is null;`.